### PR TITLE
Do not O_TRUNC writeable files before obtaining an advisory lock

### PR DESCRIPTION
**Pull request by jakeogh**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

#1066 has a implementation error, io.open() may apply the O_TRUNC flag to the underlying openat() syscall, causing the truncation of a file before an advisory lock is obtained. In the case where some other process already holds an advisory lock, O_TRUNC violates it. In the case where that other process happens to be yt-dlp, this results in a corrupt file, because the file will be truncated as the other process (the one holding the advisory lock) is writing to it.

### CODE REVIEW

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -2178,8 +2178,13 @@ class locked_file(object):
     _closed = False
 
     def __init__(self, filename, mode, block=True, encoding=None):
-        assert mode in ['r', 'rb', 'a', 'ab', 'w', 'wb']
-        self.f = io.open(filename, mode, encoding=encoding)
+        if mode not in ['r', 'rb', 'a', 'ab', 'w', 'wb']:
+            raise NotImplementedError(mode)
+        try:
+            fd = os.open(filename, os.O_WRONLY | os.O_CREAT | os.O_CLOEXEC)  # do not pass O_TRUNC
```
Comment (pukkandan) [inline yt_dlp/utils.py:2184]: where did u get these flags? They seem different from what io.open uses https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Modules/_io/fileio.c#L207

Comment (jakeogh) [reply yt_dlp/utils.py:2184]: I used strace. Typical line (before this PR) is:
```openat(AT_FDCWD</del/_yt_dlp_test_1646721829>, "youtube__Gwo3pEH7hUE.f160.mp4.part", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666)```

If there is an inconsistency regarding fileio.c, I don't see it.

**Comment by jakeogh** [comment]
Test script: https://gist.github.com/574cd581a9101c433e2c19c07a515880

**Comment by jakeogh** [comment]
Test setup: https://gist.github.com/574cd581a9101c433e2c19c07a515880

**Comment by pukkandan** [comment]
You didn't even look at what flags windows needs, did you? :(

**Pull request merged**